README.md
# Vagrant SoftLayer Provider

<span class="badges">
[![Gem Version](https://badge.fury.io/rb/vagrant-softlayer.png)](http://badge.fury.io/rb/vagrant-softlayer)
[![Code Climate](https://codeclimate.com/github/audiolize/vagrant-softlayer.png)](https://codeclimate.com/github/audiolize/vagrant-softlayer)
</span>

This is a [Vagrant](http://www.vagrantup.com) plugin that adds a [SoftLayer](http://www.softlayer.com)
provider to Vagrant, allowing Vagrant to control and provision SoftLayer CCI instances.

**NOTE:** This plugin work with Vagrant 1.3+, altough Vagrant 1.4+ is strongly advised.

## Features

* Basic lifecycle (boot, halt, reboot, suspend, resume) of SoftLayer CCI instances.
* OS reload on a CCI (`vagrant rebuild`).
* SSH into the instances.
* Provision the instances with any built-in Vagrant provisioner.
* Minimal synced folder support via `rsync`.

## Installation

> **NOTE**
>
> If you're using Vagrant 1.3, you need to [Set the SSL_CERT_FILE environment variable](https://github.com/audiolize/vagrant-softlayer/wiki/The-SSL_CERT_FILE-environment-variable) first.

Installation is performed in the prescribed manner for Vagrant 1.1+ plugins.
After installing, `vagrant up` and specify the `softlayer` provider. An example is
shown below.

```
$ vagrant plugin install vagrant-softlayer
...
$ vagrant up --provider=softlayer
...
```

Of course prior to doing this, you'll need to obtain an SoftLayer-compatible
box file for Vagrant.

## Box File Format

Every provider in Vagrant must introduce a custom box format. This
provider introduces `softlayer` boxes. You can view an example box in
the `example_box/` directory. That directory also contains instructions
on how to build a box.

The box format is basically just the required `metadata.json` file
along with a `Vagrantfile` that does default settings for the
provider-specific configuration for this provider.

## Configuration

This provider exposes quite a few provider-specific configuration options:

### Authentication

Parameter      | Description                          | Default                            | Required
-------------- | ------------------------------------ | ---------------------------------- | --------
`api_key`      | The API key for accessing SoftLayer  |                                    | yes
`endpoint_url` | The endpoint SoftLayer API url       | SoftLayer::API_PUBLIC_ENDPOINT     | yes
`username`     | The username for accessing SoftLayer |                                    | yes

> **NOTE**
>
> In place of the API key and username you can use environment variables, respectively `SL_API_KEY` and `SL_USERNAME`.

### DNS Management

If the [DNS zone](http://knowledgelayer.softlayer.com/learning/introduction-domain-name-system) of the configured domain is hosted by [SoftLayer](http://knowledgelayer.softlayer.com/topic/dns), you can automatically manage it.

Parameter    | Description                           | Default | Required
------------ | ------------------------------------- | ------- | --------
`manage_dns` | Add/remove A record on create/destroy | false   | no

### Join Local Load Balancers

See [Join load balancers](https://github.com/audiolize/vagrant-softlayer/wiki/Join-load-balancers).

### Instance Configuration

Parameter          | Description                                                | Default         | Required
------------------ | -----------------------------------------------------------| ----------------| --------
`datacenter`       | Datacenter shortname                                       | First available | no
`dedicated`        | Allocate a dedicated CCI (non-shared host)                 | false           | no
`disk_capacity`    | The capacity of each disk                                  |                 | no **
`domain`           | The domain of the instance                                 |                 | yes
`force_private_ip` | Use private IP for communication even if public IP is set  | false           | no
`hostname`         | The hostname of the instance                               |                 | yes *
`hourly_billing`   | Hourly billing type (false for monthly)                    | true            | no
`image_guid`       | The global identifier for the compute or flex image to use |                 | no **
`local_disk`       | Use a local disk (false for SAN)                           | true            | no
`max_memory`       | The amount of RAM of the instance in Mb                    | 1024            | no
`network_speed`    | Network port speed in Mbps                                 | 10              | no
`operating_system` | The instance operating system identifier                   | UBUNTU_LATEST   | no **
`post_install`     | URI of Post-install script to download                     |                 | no
`private_only`     | Only create access to the private network                  | false           | no
`ssh_key`          | ID or label of the SSH key(s) to provision                 |                 | yes
`start_cpus`       | The number of processors of the instance                   | 1               | no
`user_data`        | User defined metadata string                               |                 | no
`vlan_private`     | The ID, name or qualified name of the private VLAN         | Auto-generated  | no
`vlan_public`      | The ID, name or qualified name of the public VLAN          | Auto-generated  | no

\* The `hostname` could be specified either using `config.vm.hostname` or the provider parameter.

\** When defining the instance you can either specify an `image_guid` or `operating_system` with optional `disk_capacity`.

### Timeouts/Waits

Parameter           | Description                               | Default         | Required
------------------- | ------------------------------------------| ----------------| --------
`api_timeout`       | SoftLayer API call timeout in seconds     | 60              | no
`provision_timeout` | Provisioning wait timeout in seconds      | 1200            | no
`rebuild_timeout`   | Rebuild wait timeout in seconds           | 1200            | no
`transaction_wait`  | Wait for order transactions to complete   | true            | no

These can be set like typical provider-specific configuration:

```
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :softlayer do |sl|
    sl.api_key  = "foo"
    sl.username = "bar"
    sl.ssh_key  = "Vagrant insecure key"
  end
end
```

## OS Reload

Reload of an instance's operating system is performed with the `vagrant rebuild` command.
The primary disk of the instance will be formatted and a fresh copy of the underlying OS
will be applied. Note that only `ssh_key` and `post_install` parameter will be read
during rebuild, all the other parameters will be ignored. Provisioners will always run
after rebuild.

## Synced Folders

There is minimal support for synced folders. Upon `vagrant up`,
`vagrant reload`, and `vagrant provision`, the SoftLayer provider will use
`rsync` (if available) to uni-directionally sync the folder to
the remote machine over SSH.

This is good enough for all built-in Vagrant provisioners (shell, ansible,
chef, and puppet) to work!

## Other Examples

### Multiple disks

The `disk_capacity` parameter accepts an hash with the following structure:

```
{
  disk_id => disk_size,
  disk_id => disk_size,
  ...
}
```

Disk ID 1 is reserved is reserved for swap space, in the following example two disks
of 25Gb and 100Gb will be provisioned:

```
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :softlayer do |sl, override|
    # ... other stuff

    sl.disk_capacity = { 0 => 25, 2 => 100 }
  end
end
```

### Override SSH Username

If you're running Vagrant with an user different from root, probably you need
to override the username used for ssh connection. You can do it using the standard
Vagrant syntax:

```
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :softlayer do |sl, override|
    # ... other stuff

    override.ssh.username = "root"
  end
end
```

### Multiple SSH keys

Multiple SSH keys to be provisioned could be specified using an array:

```
Vagrant.configure("2") do |config|
  # ... other stuff

  config.vm.provider :softlayer do |sl|
    sl.api_key  = "foo"
    sl.username = "bar"
    # ssh_keys is just an alias of ssh_key
    sl.ssh_keys = ["Vagrant insecure key", "My personal key"]
  end
end
```

Also, a bunch of aliases for the `ssh_key` parameter are provided for better semantic:

* `ssh_keys`
* `ssh_key_id`
* `ssh_key_ids`
* `ssh_key_name`
* `ssh_key_names`

## Quick Start Guide

For those new to Vagrant, see the [Quick Start Guide](https://github.com/audiolize/vagrant-softlayer/blob/master/QUICKSTART.md).

## Development

To work on the `vagrant-softlayer` plugin, clone this repository out, and use
[Bundler](http://gembundler.com) to get the dependencies:

```
$ bundle
```

Once you have the dependencies, verify the unit tests pass with `rake`:

```
$ bundle exec rake
```

If those pass, you're ready to start developing the plugin. You can test
the plugin without installing it into your Vagrant environment by just
creating a `Vagrantfile` in the top level of this directory (it is gitignored)
and add the following line to your `Vagrantfile`:

```
Vagrant.require_plugin "vagrant-softlayer"
```

Use bundler to execute Vagrant:

```
$ bundle exec vagrant up --provider=softlayer
```

## Contributing

`vagrant-softlayer` is following [git-flow](https://github.com/nvie/gitflow) branch/workflow management. Please
submit any pull requests to the `develop` branch.

## Credits

Emiliano Ticci (@emyl)
Julio Lajara (@ju2wheels)


CHANGELOG.md
## 0.5.0 (UNRELEASED)

## 0.4.1 (March 5, 2015)
BUG FIXES:
  - Fix syntax errors in spec and pin rpsec versioning
  - Fix incorrect filter on VLANs contrib tool causing it to not list all available VLANs
  - Pin softlayer_api version to >= 3.0.1 to fix bug in wait_until_ready

## 0.4.0 (October 1, 2014)
ENHANCEMENTS:
  - Begins transition to SoftLayer API 2.x/3.x
  - Add support for SoftLayer API timeout

## 0.3.4 (October 1, 2014)
ENHANCEMENTS:
  - Add transaction_wait provider option.

BUG FIXES:
  - Backport SoftLayer::VirtualServer#wait_until_ready to fix transaction wait problems.

## 0.3.3 (August 21, 2014)

ENHANCEMENTS:

  - Add parallel execution capability.
  - Add configuration variables for provision and rebuild timeout.
  - Use builtin synced folders if available.

BUG FIXES:

  - Raise a proper exception on timeout.

## 0.3.2 (August 1, 2014)

ENHANCEMENTS:

  - Add ability to specify vlans by name or qualified name.
  - Add vagrant-softlayer-vlans contrib tool.
  - Add ability to force private IP address usage.
  - Add quick start guide and Vagrant Cloud boxes.

BUG FIXES:

  - Fix crash on destroy/rebuild when the instance don't exist.

## 0.3.1 (June 3, 2014)

BUG FIXES:

  - Lock down softlayer_api gem to 1.0.x

## 0.3.0 (April 15, 2014)

NEW FEATURES:

  - Add selection of disk sizes for templates.
  - Add selection of block devices by id.
  - Add vagrant-softlayer-boxes contrib tool for creating boxes.
  - Add suspend/resume commands.

## 0.2.0 (December 5, 2013)

NEW FEATURES:

  - Automatic joining of local load balancers.

## 0.1.0 (November 18, 2013)

  - Initial release.


lib/vagrant-softlayer/config.rb
require "ostruct"

module VagrantPlugins
  module SoftLayer
    class Config < Vagrant.plugin("2", :config)
      # The API key to access SoftLayer.
      attr_accessor :api_key

      # The endpoint SoftLayer API url.
      attr_accessor :endpoint_url

      # The username to access SoftLayer.
      attr_accessor :username

      # The datacenter shortname.
      attr_accessor :datacenter

      # Whether to allocate a dedicated instance.
      attr_accessor :dedicated

      # The disk image capacity
      attr_accessor :disk_capacity

      # The domain of the instance.
      attr_accessor :domain

      # Force the use of the private IP for all communication even if a public IP is available
      attr_accessor :force_private_ip

      # The hostname of the instance.
      attr_accessor :hostname

      # The billing type of the instance (true for hourly, false for monthly).
      attr_accessor :hourly_billing

      # The global identifier of the compute or flex image to use.
      attr_accessor :image_guid

      # The disk type of the instance (true for local, false for SAN).
      attr_accessor :local_disk

      # The amount of RAM of the instance.
      attr_accessor :max_memory

      # Network port speed in Mbps.
      attr_accessor :network_speed

      # The instance operating system identifier.
      attr_accessor :operating_system

      # URI of post-install script to download.
      attr_accessor :post_install

      # Whether or not the instance only has access to the private network.
      attr_accessor :private_only

      # The amount of time in seconds to wait for provision to complete.
      attr_accessor :provision_timeout

      # The amount of time in seconds to wait for rebuild to complete.
      attr_accessor :rebuild_timeout

      # The id or name of the ssh key to be provisioned.
      attr_accessor :ssh_key

      # The number of processors of the instance.
      attr_accessor :start_cpus

      # Whether to wait for transaction completion on actions that place orders
      attr_accessor :transaction_wait

      # User defined metadata string.
      attr_accessor :user_data

      # The ID, name or qualified name of the private VLAN.
      attr_accessor :vlan_private

      # The ID, name or qualified name of the public VLAN.
      attr_accessor :vlan_public

      # The load balancers service groups to join.
      attr_reader :load_balancers

      # Automatically update DNS on create and destroy.
      attr_accessor :manage_dns

      def initialize
        @api_key      = UNSET_VALUE
        @endpoint_url = UNSET_VALUE
        @username     = UNSET_VALUE

        @datacenter        = UNSET_VALUE
        @dedicated         = UNSET_VALUE
        @disk_capacity     = UNSET_VALUE
        @domain            = UNSET_VALUE
        @force_private_ip  = UNSET_VALUE
        @hostname          = UNSET_VALUE
        @image_guid        = UNSET_VALUE
        @hourly_billing    = UNSET_VALUE
        @local_disk        = UNSET_VALUE
        @max_memory        = UNSET_VALUE
        @network_speed     = UNSET_VALUE
        @operating_system  = UNSET_VALUE
        @post_install      = UNSET_VALUE
        @private_only      = UNSET_VALUE
        @provision_timeout = UNSET_VALUE
        @rebuild_timeout   = UNSET_VALUE
        @ssh_key           = UNSET_VALUE
        @start_cpus        = UNSET_VALUE
        @transaction_wait  = UNSET_VALUE
        @user_data         = UNSET_VALUE
        @vlan_private      = UNSET_VALUE
        @vlan_public       = UNSET_VALUE

        @load_balancers = []
        @manage_dns     = UNSET_VALUE
      end

      # Set the load balancer service group to join.
      #
      # Available options:
      #
      # :method => Routing method. Default to round robin.
      # :port   => Load balancer virtual port.
      # :type   => Routing type. Default to TCP.
      # :vip    => Load balancer virtual IP address.
      #
      # An optional block will accept parameters for the
      # balanced service. Available parameters:
      #
      # :destination_port => TCP port on the node.
      # :health_check     => Service health check. Default to ping.
      # :weight           => Service weight. Default to 1.
      #
      def join_load_balancer(opts = {}, &block)
        # Defaults
        opts[:method] ||= "ROUND ROBIN"
        opts[:type]   ||= "TCP"
        opts[:service]  = OpenStruct.new(:destination_port => nil, :health_check => "PING", :weight => 1)

        yield opts[:service] if block_given?

        # Convert all options that belongs to
        # an enumeration in uppercase.
        opts[:method].upcase!
        opts[:type].upcase!
        opts[:service].health_check.upcase!

        @load_balancers << opts
      end

      def finalize!
        # Try to get username and api key from environment variables.
        # They will default to nil if the environment variables are not present.
        @api_key  = ENV["SL_API_KEY"] if @api_key == UNSET_VALUE
        @username = ENV["SL_USERNAME"] if @username == UNSET_VALUE

        # Endpoint url defaults to public SoftLayer API url.
        @endpoint_url = API_PUBLIC_ENDPOINT if @endpoint_url == UNSET_VALUE

        # No default datacenter.
        @datacenter = nil if @datacenter == UNSET_VALUE

        # Shared instance by default.
        @dedicated = false if @dedicated == UNSET_VALUE

        # 25GB disk capacity image by default.
        @disk_capacity = nil if @disk_capacity == UNSET_VALUE

        # Domain should be specified in Vagrantfile, so we set default to nil.
        @domain = nil if @domain == UNSET_VALUE

        # Disable the use of force private IP so the default selection can take effect
        @force_private_ip = false if @force_private_ip == UNSET_VALUE

        # Hostname should be specified in Vagrantfile, either using `config.vm.hostname`
        # or the provider specific configuration entry.
        @hostname = nil if @hostname == UNSET_VALUE

        # Bill hourly by default.
        @hourly_billing = true if @hourly_billing == UNSET_VALUE

        # Disable the use of a specific block device image so we can leave the OS template as default
        @image_guid = nil if @image_guid == UNSET_VALUE

        # Use local disk by default.
        @local_disk = true if @local_disk == UNSET_VALUE

        # 1Gb of RAM by default.
        @max_memory = 1024 if @max_memory == UNSET_VALUE

        # 10Mbps by default.
        @network_speed = 10 if @network_speed == UNSET_VALUE

        # Provision with the latest Ubuntu by default.
        @operating_system = "UBUNTU_LATEST" if @operating_system == UNSET_VALUE

        # No post install script by default.
        @post_install = nil if @post_install == UNSET_VALUE

        # Private-network only is false by default.
        @private_only = false if @private_only == UNSET_VALUE

        # The amount of time in seconds to wait for provision to complete.
        @provision_timeout = 1200 if @provision_timeout == UNSET_VALUE

        # The amount of time in seconds to wait for rebuild to complete.
        @rebuild_timeout = 1200 if @rebuild_timeout == UNSET_VALUE

        # SSH key should be specified in Vagrantfile, so we set default to nil.
        @ssh_key = nil if @ssh_key == UNSET_VALUE

        # One processor by default.
        @start_cpus = 1 if @start_cpus == UNSET_VALUE

        # Whether to wait for transaction completion on actions that place orders
        @transaction_wait = true if @start_cpus == UNSET_VALUE

        # No user metadata by default.
        @user_data = nil if @user_data == UNSET_VALUE

        # No specific private VLAN by default.
        @vlan_private = nil if @vlan_private == UNSET_VALUE

        # No specific public VLAN by default.
        @vlan_public = nil if @vlan_public == UNSET_VALUE

        # DNS management off by default
        @manage_dns = false if @manage_dns == UNSET_VALUE
      end

      # Aliases for ssh_key for beautiful semantic.
      def ssh_keys=(value)
        @ssh_key = value
      end

      alias_method :ssh_key_id=, :ssh_keys=
      alias_method :ssh_key_ids=, :ssh_keys=
      alias_method :ssh_key_name=, :ssh_keys=
      alias_method :ssh_key_names=, :ssh_keys=

      def validate(machine)
        errors = []

        errors << I18n.t("vagrant_softlayer.config.api_key_required") if !@api_key
        errors << I18n.t("vagrant_softlayer.config.username_required") if !@username

        errors << I18n.t("vagrant_softlayer.config.domain_required") if !@domain
        errors << I18n.t("vagrant_softlayer.config.ssh_key_required") if !@ssh_key

        errors << I18n.t("vagrant_softlayer.config.img_guid_os_code_mutually_exclusive") if @image_guid && @operating_system
        errors << I18n.t("vagrant_softlayer.config.img_guid_capacity_mutually_exclusive") if @image_guid && @disk_capacity

        #  Fail if both `vm.hostname` and `provider.hostname` are nil.
        if !@hostname && !machine.config.vm.hostname
          errors << I18n.t("vagrant_softlayer.config.hostname_required")
        end

        # Fail if a load balancer has been specified without vip, port or destination port.
        unless @load_balancers.empty?
          @load_balancers.each do |lb|
            errors << I18n.t("vagrant_softlayer.config.lb_port_vip_required") unless (lb[:vip] && lb[:port] && lb[:service].destination_port)
          end
        end

        # Fail if two or more load balancers has been specified with same vip and port.
        if @load_balancers.map { |lb| { :port => lb[:port], :vip => lb[:vip] } }.uniq!
          errors << I18n.t("vagrant_softlayer.config.lb_duplicate")
        end

        { "SoftLayer" => errors }
      end
    end
  end
end
